syntax = "proto3";

package ai.h2o.storage;

import "api/v1/proto/listing.proto";
import "api/v1/proto/user.proto";

option go_package = "github.com/h2oai/h2oai-storage/api/v1/proto";
option java_multiple_files = true;
option java_package = "ai.h2o.storage";

service UserService {
  rpc GetUser(GetUserRequest) returns (GetUserResponse) {}
  rpc BatchGetUsers(BatchGetUsersRequest) returns (BatchGetUsersResponse) {}
  rpc GetUserByUsername(GetUserByUsernameRequest)
      returns (GetUserByUsernameResponse) {}
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse) {}
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse) {}
  rpc WhoAmI(WhoAmIRequest) returns (WhoAmIResponse) {}
}

message GetUserRequest { string id = 1; }

message GetUserResponse { User user = 1; }

message BatchGetUsersRequest { repeated string IDs = 1;}

message BatchGetUsersResponse {
  repeated User users = 1;
}

message GetUserByUsernameRequest { string username = 1; }

message GetUserByUsernameResponse { User user = 1; }

message CreateUserRequest { User user = 1; }

message CreateUserResponse { User user = 1; }

message ListUsersRequest {
  PagingRequest paging = 1;
  FilterRequest filter = 2;
  SortingRequest sorting = 3;
}

message ListUsersResponse {
  PagingResponse paging = 1;
  repeated User user = 2;
}

message WhoAmIRequest {}

message WhoAmIResponse { User user = 1; }
